File: src/nfan/fan/sys/Map.cs
//
// Copyright (c) 2006, Brian Frank and Andy Frank
// Licensed under the Academic Free License version 3.0
//
// History:
//   28 Sep 06  Andy Frank  Creation
//

using System;
using System.Collections;
using System.Diagnostics;
using System.Text;
using Fanx.Serial;

namespace Fan.Sys
{
  /// <summary>
  /// Map is a hashm_map of key value pairs.
  /// </summary>
  public sealed class Map : FanObj, Literal
  {

  //////////////////////////////////////////////////////////////////////////
  // Constructors
  //////////////////////////////////////////////////////////////////////////

    public static Map make(Type type)
    {
      return new Map((MapType)type, new Hashtable());
    }

    public Map(Type k, Type v) : this(new MapType(k, v), new Hashtable())
    {
    }

    public Map(MapType type) : this(type, new Hashtable())
    {
    }

    public Map(MapType type, Hashtable map)
    {
      if (type == null || map == null)
      {
        Console.WriteLine(new StackTrace(true));
        throw new NullErr().val;
      }

      this.m_type = type;
      this.m_map  = map;
    }

  //////////////////////////////////////////////////////////////////////////
  // Identity
  //////////////////////////////////////////////////////////////////////////

    public override Type type()
    {
      return m_type;
    }

  //////////////////////////////////////////////////////////////////////////
  // Methods
  //////////////////////////////////////////////////////////////////////////

    public bool isEmpty()
    {
      return m_map.Count == 0;
    }

    public long size()
    {
      return m_map.Count;
    }

    public object get(object key)
    {
      if (key == null) return m_def;
      object val = m_map[key];
      if (val != null) return val;
      return m_def;
    }

    public object get(object key, object def)
    {
      if (key == null) return def;
      object val = m_map[key];
      if (val != null) return val;
      return def;
    }

    public bool containsKey(object key)
    {
      return (key == null) ? false : m_map.ContainsKey(key);
    }

    public List keys()
    {
      return new List(m_type.m_k, m_map.Keys);
    }

    public List values()
    {
      return new List(m_type.m_v, m_map.Values);
    }

    public Map set(object key, object val)
    {
      modify();
      if (key == null)
        throw NullErr.make("key is null").val;
      if (!isImmutable(key))
        throw NotImmutableErr.make("key is not immutable: " + type(key)).val;
      m_map[key] = val;
      return this;
    }

    public Map add(object key, object val)
    {
      modify();
      if (key == null)
        throw NullErr.make("key is null").val;
      if (!isImmutable(key))
        throw NotImmutableErr.make("key is not immutable: " + type(key)).val;
      if (m_map[key] != null)
        throw ArgErr.make("Key already mapped: " + key).val;
      m_map[key] = val;
      return this;
    }

    public Map setAll(Map m)
    {
      modify();
      IDictionaryEnumerator en = m.m_map.GetEnumerator();
      while (en.MoveNext())
      {
        object key = en.Key;
        object val = en.Value;
        m_map[key] = val;
      }
      return this;
    }

    public Map addAll(Map m)
    {
      modify();
      IDictionaryEnumerator en = m.m_map.GetEnumerator();
      while (en.MoveNext())
      {
        object key = en.Key;
        object val = en.Value;
        add(key, val);
      }
      return this;
    }

    public object remove(object key)
    {
      modify();
      object val = m_map[key];
      m_map.Remove(key);
      return val;
    }

    public Map dup()
    {
      Map dup = new Map(m_type);
      dup.m_map = (Hashtable)this.m_map.Clone();
      return dup;
    }

    public void clear()
    {
      modify();
      m_map.Clear();
    }

    public bool caseInsensitive() { return m_caseInsensitive; }
    public void caseInsensitive(bool v)
    {
      modify();

      if (m_type.m_k != Sys.StrType)
        throw UnsupportedErr.make("Map not keyed by string: " + m_type).val;

      if (m_map.Count != 0)
        throw UnsupportedErr.make("Map not empty").val;

      if (this.m_caseInsensitive == v) return;
      this.m_caseInsensitive = v;

      if (m_caseInsensitive)
        m_map = new Hashtable(new CIEqualityComparer());
      else
        m_map = new Hashtable();
    }

    public bool ordered()
    {
      throw UnsupportedErr.make("Not yet implemented").val;
    }

    public void ordered(bool v)
    {
      // TODO
      // http://hg.fandev.org/repos/fan-1.0?cs=bc1b2c7e511f

      throw UnsupportedErr.make("Not yet implemented").val;
    }

    public object def() { return m_def; }
    public void def(object v)
    {
      modify();
      if (v != null && !isImmutable(v))
        throw NotImmutableErr.make("def must be immutable: " + type(v)).val;
      this.m_def = v;
    }

    public override bool Equals(object that)
    {
      if (that is Map)
      {
        if (!m_type.Equals(type(that)))
          return false;

        Hashtable thatMap = ((Map)that).m_map;
        if (m_map.Count != thatMap.Count)
          return false;

        IDictionaryEnumerator en = m_map.GetEnumerator();
        while (en.MoveNext())
        {
          object key  = en.Key;
          object val  = en.Value;
          object test = thatMap[key];

          if (val == null)
          {
            if (test != null) return false;
          }
          else if (!val.Equals(test))
          {
            return false;
          }
        }

        return true;
      }
      return false;
    }

    public override int GetHashCode() { return (int)hash(); }

    public override long hash()
    {
      int hash = 0;
      IDictionaryEnumerator en = m_map.GetEnumerator();
      while (en.MoveNext())
      {
        object key = en.Key;
        object val = en.Value;
        hash += key.GetHashCode() ^ (val == null ? 0 : val.GetHashCode());
      }
      return hash;
    }

    public override string toStr()
    {
      if (m_map.Count == 0) return "[:]";
      StringBuilder s = new StringBuilder(32+m_map.Count*32);
      s.Append("[");
      bool first = true;
      IDictionaryEnumerator en = m_map.GetEnumerator();
      while (en.MoveNext())
      {
        object key = en.Key;
        object val = en.Value;
        if (!first) s.Append(", ");
        else first = false;
        s.Append(key).Append(':').Append(val);
      }
      s.Append("]");
      return s.ToString();
    }

    public void encode(ObjEncoder @out)
    {
      // route back to obj encoder
      @out.writeMap(this);
    }

  //////////////////////////////////////////////////////////////////////////
  // Iterators
  //////////////////////////////////////////////////////////////////////////

    public void each(Func f)
    {
      IDictionaryEnumerator en = m_map.GetEnumerator();
      while (en.MoveNext())
      {
        object key = en.Key;
        object val = en.Value;
        f.call2(val, key);
      }
    }

    public object eachWhile(Func f)
    {
      IDictionaryEnumerator en = m_map.GetEnumerator();
      while (en.MoveNext())
      {
        object key = en.Key;
        object val = en.Value;
        object r = f.call2(val, key);
        if (r != null) return r;
      }
      return null;
    }

    public object find(Func f)
    {
      IDictionaryEnumerator en = m_map.GetEnumerator();
      while (en.MoveNext())
      {
        object key = en.Key;
        object val = en.Value;
        if (f.call2(val, key) == Boolean.True)
          return val;
      }
      return null;
    }

    public Map findAll(Func f)
    {
      Map acc = new Map(m_type);
      IDictionaryEnumerator en = m_map.GetEnumerator();
      while (en.MoveNext())
      {
        object key = en.Key;
        object val = en.Value;
        if (f.call2(val, key) == Boolean.True)
          acc.set(key, val);
      }
      return acc;
    }

    public Map exclude(Func f)
    {
      Map acc = new Map(m_type);
      IDictionaryEnumerator en = m_map.GetEnumerator();
      while (en.MoveNext())
      {
        object key = en.Key;
        object val = en.Value;
        if (f.call2(val, key) == Boolean.False)
          acc.set(key, val);
      }
      return acc;
    }

    public object reduce(object reduction, Func f)
    {
      IDictionaryEnumerator en = m_map.GetEnumerator();
      while (en.MoveNext())
      {
        object key = en.Key;
        object val = en.Value;
        reduction = f.call3(reduction, val, key);
      }
      return reduction;
    }

    public Map map(Map acc, Func f)
    {
      IDictionaryEnumerator en = m_map.GetEnumerator();
      while (en.MoveNext())
      {
        object key = en.Key;
        object val = en.Value;
        acc.set(key, f.call2(val, key));
      }
      return acc;
    }

  //////////////////////////////////////////////////////////////////////////
  // Join
  //////////////////////////////////////////////////////////////////////////

    public string join(string sep) { return join(sep, null); }
    public string join(string sep, Func f)
    {
      int size = (int)this.size();
      if (size == 0) return "";
      StringBuilder s = new StringBuilder(32+size*32);
      IDictionaryEnumerator en = m_map.GetEnumerator();
      while (en.MoveNext())
      {
        object key = en.Key;
        object val = en.Value;
        if (s.Length > 0) s.Append(sep);
        if (f == null)
          s.Append(key).Append(": ").Append(val);
        else
          s.Append(f.call2(val, key));
      }
      return s.ToString();
    }

  //////////////////////////////////////////////////////////////////////////
  // Readonly
  //////////////////////////////////////////////////////////////////////////

    public bool isRW()
    {
      return !m_isReadonly;
    }

    public bool isRO()
    {
      return m_isReadonly;
    }

    public Map rw()
    {
      if (!m_isReadonly) return this;

      Map rw = new Map(m_type);
      rw.m_map = (Hashtable)m_map.Clone();
      rw.m_isReadonly = false;
      rw.m_readonlyMap = this;
      rw.m_caseInsensitive = m_caseInsensitive;
      rw.m_def = m_def;
      return rw;
    }

    public Map ro()
    {
      if (m_isReadonly) return this;
      if (m_readonlyMap == null)
      {
        Map ro = new Map(m_type);
        ro.m_map = m_map;
        ro.m_isReadonly  = true;
        ro.m_caseInsensitive = m_caseInsensitive;
        ro.m_def = m_def;
        m_readonlyMap = ro;
      }
      return m_readonlyMap;
    }

    public override bool isImmutable()
    {
      return m_immutable;
    }

    public Map toImmutable()
    {
      if (m_immutable) return this;

      // make safe copy
      Hashtable temp = m_caseInsensitive
        ? new Hashtable(new CIEqualityComparer()) : new Hashtable();
      IDictionaryEnumerator en = m_map.GetEnumerator();
      while (en.MoveNext())
      {
        object key = en.Key;
        object val = en.Value;

        if (val != null)
        {
          if (val is List)
            val = ((List)val).toImmutable();
          else if (val is Map)
            val = ((Map)val).toImmutable();
          else if (!isImmutable(val))
            throw NotImmutableErr.make("Item [" + key + "] not immutable " + type(val)).val;
        }

        temp[key] = val;
      }

      // return new m_immutable m_map
      Map ro = new Map(m_type, temp);
      ro.m_isReadonly = true;
      ro.m_immutable = true;
      ro.m_caseInsensitive = m_caseInsensitive;
      ro.m_def = m_def;
      return ro;
    }

    private void modify()
    {
      // if readonly then throw readonly exception
      if (m_isReadonly)
        throw ReadonlyErr.make("Map is readonly").val;

      // if we have a cached m_readonlyMap, then detach
      // it so it remains m_immutable
      if (m_readonlyMap != null)
      {
        m_readonlyMap.m_map = (Hashtable)m_map.Clone();
        m_readonlyMap = null;
      }
    }

  //////////////////////////////////////////////////////////////////////////
  // C#
  //////////////////////////////////////////////////////////////////////////

    public IDictionaryEnumerator pairsIterator()
    {
      return m_map.GetEnumerator();
    }

    public IEnumerator keysEnumerator()
    {
      return m_map.Keys.GetEnumerator();
    }

  //////////////////////////////////////////////////////////////////////////
  // CIEqualityComparer (Case Insensitive)
  //////////////////////////////////////////////////////////////////////////

    class CIEqualityComparer : IEqualityComparer
    {
      public new bool Equals(object x, object y)
      {
        return FanStr.equalsIgnoreCase((string)x, (string)y);
      }

      public int GetHashCode(object obj)
      {
        return FanStr.caseInsensitiveHash((string)obj);
      }
    }

  //////////////////////////////////////////////////////////////////////////
  // Fields
  //////////////////////////////////////////////////////////////////////////

    private MapType m_type;
    private Hashtable m_map;
    private Map m_readonlyMap;
    private bool m_isReadonly;
    private bool m_immutable;
    private bool m_caseInsensitive = false;
    private object m_def;

  }
}

File: src/sys/java/fan/sys/Map.java
//
// Copyright (c) 2006, Brian Frank and Andy Frank
// Licensed under the Academic Free License version 3.0
//
// History:
//   17 Mar 06  Brian Frank  Creation
//
package fan.sys;

import java.lang.Thread;
import java.util.AbstractSet;
import java.util.Arrays;
import java.util.Collection;
import java.util.HashMap;
import java.util.Map.Entry;
import java.util.LinkedHashMap;
import java.util.Iterator;
import java.util.Set;
import fanx.serial.*;
import fanx.util.OpUtil;

/**
 * Map is a hashmap of key value pairs.
 */
public final class Map<K,V>
  extends FanObj
  implements Literal, java.util.Map<K,V>
{

//////////////////////////////////////////////////////////////////////////
// Constructors
//////////////////////////////////////////////////////////////////////////

  public static Map make(Type type)
  {
    MapType t = null;
    try
    {
      t = (MapType)type;
    }
    catch (ClassCastException e)
    {
      throw ArgErr.make("Non-nullable map type required: " + type);
    }
    if (t.k.isNullable()) throw ArgErr.make("Map key type cannot be nullable: " + t.k);
    return new Map(t, new HashMap());
  }

  /** Construct map with given key and value types */
  public static Map make(Type k, Type v)
  {
    return new Map(k, v);
  }

  Map(Type k, Type v)
  {
    this(new MapType(k, v), new HashMap());
  }

  Map(MapType type)
  {
    this(type, new HashMap());
  }

  Map(MapType type, HashMap map)
  {
    if (type == null || map == null) { Thread.dumpStack(); throw NullErr.make(); }
    this.type = type;
    this.map  = map;
  }

//////////////////////////////////////////////////////////////////////////
// Identity
//////////////////////////////////////////////////////////////////////////

  public final Type typeof()
  {
    return type;
  }

  MapType type()
  {
    return type;
  }

//////////////////////////////////////////////////////////////////////////
// Methods
//////////////////////////////////////////////////////////////////////////

  public final boolean isEmpty()
  {
    return map.size() == 0;
  }

  public final long _size()
  {
    return map.size();
  }

  public final V get(Object key)
  {
    V val = map.get(key);
    if (val != null) return val;
    if (this.def == null) return null;
    return map.containsKey(key) ? null : this.def;
  }

  public final V get(K key, V def)
  {
    V val = map.get(key);
    if (val != null) return val;
    if (def == null) return null;
    return map.containsKey(key) ? null : def;
  }

  public final V getChecked(K key) { return getChecked(key, true); }
  public final V getChecked(K key, boolean checked)
  {
    V val = map.get(key);
    if (val != null) return val;
    if (map.containsKey(key)) return null;
    if (checked) throw UnknownKeyErr.make(String.valueOf(key));
    return null;
  }

  public final V getOrThrow(K key)
  {
    V val = map.get(key);
    if (val != null) return val;
    if (map.containsKey(key)) return null;
    throw UnknownKeyErr.make(String.valueOf(key));
  }

  public final boolean containsKey(Object key)
  {
    return map.containsKey(key);
  }

  public final List<K> keys()
  {
    Object[] keys = new Object[map.size()];
    Iterator it = pairsIterator();
    for (int i=0; it.hasNext(); ++i)
      keys[i] = ((Entry)it.next()).getKey();
    return new List(type.k, keys);
  }

  public final List<V> vals()
  {
    return new List(type.v, map.values());
  }

  public final Map<K,V> set(K key, V value)
  {
    modify();
    if (key == null)
      throw NullErr.make("key is null");
    if (!isImmutable(key))
      throw NotImmutableErr.make("key is not immutable: " + typeof(key));
    map.put(key, value);
    return this;
  }

  public final Map<K,V> setNotNull(K key, V value)
  {
    if (value == null) return this;
    return set(key, value);
  }

  public final Map<K,V> add(K key, V value)
  {
    modify();
    if (key == null)
      throw NullErr.make("key is null");
    if (!isImmutable(key))
      throw NotImmutableErr.make("key is not immutable: " + typeof(key));
    if (map.containsKey(key))
      throw ArgErr.make("Key already mapped: " + key);
    map.put(key, value);
    return this;
  }

  public final Map<K,V> addIfNotNull(K key, V value)
  {
    return addNotNull(key, value);
  }

  public final Map<K,V> addNotNull(K key, V value)
  {
    if (value == null) return this;
    return add(key, value);
  }

  public final V getOrAdd(K key, Func valFunc)
  {
    if (map.containsKey(key)) return map.get(key);
    V val = (V)valFunc.call(key);
    add(key, val);
    return val;
  }

  public final Map<K,V> setAll(Map<K,V> m)
  {
    modify();
    Iterator it = m.pairsIterator();
    while (it.hasNext())
    {
      Entry<K,V> e = (Entry)it.next();
      map.put(e.getKey(), e.getValue());
    }
    return this;
  }

  public final Map<K,V> addAll(Map<K,V> m)
  {
    modify();
    Iterator it = m.pairsIterator();
    while (it.hasNext())
    {
      Entry<K,V> e = (Entry)it.next();
      add(e.getKey(), e.getValue());
    }
    return this;
  }

  public final Map<K,V> setList(List<V> list) { return setList(list, null); }
  public final Map<K,V> setList(List<V> list, Func f)
  {
    modify();
    if (f == null)
    {
      for (int i=0; i<list.sz(); ++i)
        set((K)list.get(i), list.get(i));
    }
    else if (f.arity() == 1)
    {
      for (int i=0; i<list.sz(); ++i)
        set((K)f.call(list.get(i)), list.get(i));
    }
    else
    {
      for (int i=0; i<list.sz(); ++i)
        set((K)f.call(list.get(i), Long.valueOf(i)), list.get(i));
    }
    return this;
  }

  public final Map<K,V> addList(List<V> list) { return addList(list, null); }
  public final Map<K,V> addList(List<V> list, Func f)
  {
    modify();
    if (f == null)
    {
      for (int i=0; i<list.sz(); ++i)
        add((K)list.get(i), list.get(i));
    }
    else if (f.arity() == 1)
    {
      for (int i=0; i<list.sz(); ++i)
        add((K)f.call(list.get(i)), list.get(i));
    }
    else
    {
      for (int i=0; i<list.sz(); ++i)
        add((K)f.call(list.get(i), Long.valueOf(i)), list.get(i));
    }
    return this;
  }

  public final V remove(Object key)
  {
    modify();
    return map.remove(key);
  }

  public final Map<K,V> dup()
  {
    Map dup = new Map(type);
    dup.map = (HashMap)this.map.clone();
    return dup;
  }

  public final Map<K,V> _clear()
  {
    modify();
    map.clear();
    return this;
  }

  public final boolean caseInsensitive()
  {
    return map instanceof CIHashMap;
  }

  public final void caseInsensitive(boolean v)
  {
    modify();

    if (type.k != Sys.StrType)
      throw UnsupportedErr.make("Map not keyed by Str: " + type);

    if (map.size() != 0)
      throw UnsupportedErr.make("Map not empty");

    if (v && ordered())
      throw UnsupportedErr.make("Map cannot be caseInsensitive and ordered");

    if (caseInsensitive() == v) return;

    if (v)
      map = new CIHashMap();
    else
      map = new HashMap();
  }

  public final boolean ordered()
  {
    return map instanceof LinkedHashMap;
  }

  public final void ordered(boolean v)
  {
    modify();

    if (map.size() != 0)
      throw UnsupportedErr.make("Map not empty");

    if (v && caseInsensitive())
      throw UnsupportedErr.make("Map cannot be caseInsensitive and ordered");

    if (ordered() == v) return;

    if (v)
      map = new LinkedHashMap();
    else
      map = new HashMap();
  }

  public final V def() { return def; }
  public final void def(V v)
  {
    modify();
    if (v != null && !isImmutable(v))
      throw NotImmutableErr.make("def must be immutable: " + typeof(v));
    this.def = v;
  }

  public final boolean equals(Object that)
  {
    if (that instanceof Map)
    {
      return type.equals(typeof(that)) && map.equals(((Map)that).map);
    }
    return false;
  }

  public final long hash()
  {
    return map.hashCode();
  }

  public final String toStr()
  {
    if (map.size() == 0) return "[:]";
    StringBuilder s = new StringBuilder(32+map.size()*32);
    s.append("[");
    boolean first = true;
    Iterator it = pairsIterator();
    while (it.hasNext())
    {
      Entry e = (Entry)it.next();
      if (!first) s.append(", ");
      else first = false;
      s.append(e.getKey()).append(':').append(e.getValue());
    }
    s.append("]");
    return s.toString();
  }

  public final void encode(ObjEncoder out)
  {
    // route back to obj encoder
    out.writeMap(this);
  }

//////////////////////////////////////////////////////////////////////////
// Iterators
//////////////////////////////////////////////////////////////////////////

  public final void each(Func f)
  {
    Iterator it = pairsIterator();
    while (it.hasNext())
    {
      Entry e = (Entry)it.next();
      f.call(e.getValue(), e.getKey());
    }
  }

  public final Object eachWhile(Func f)
  {
    Iterator it = pairsIterator();
    while (it.hasNext())
    {
      Entry e = (Entry)it.next();
      Object r = f.call(e.getValue(), e.getKey());
      if (r != null) return r;
    }
    return null;
  }

  public final V find(Func f)
  {
    Iterator it = pairsIterator();
    while (it.hasNext())
    {
      Entry<K,V> e = (Entry)it.next();
      K key = e.getKey();
      V val = e.getValue();
      if (f.callBool(val, key))
        return val;
    }
    return null;
  }

  public final Map<K,V> findAll(Func f)
  {
    Map acc = new Map(type);
    if (this.ordered()) acc.ordered(true);
    if (this.caseInsensitive()) acc.caseInsensitive(true);
    Iterator it = pairsIterator();
    while (it.hasNext())
    {
      Entry e = (Entry)it.next();
      Object key = e.getKey();
      Object val = e.getValue();
      if (f.callBool(val, key))
        acc.set(key, val);
    }
    return acc;
  }

  public final Map<K,V> findNotNull()
  {
    Map acc = new Map(type.k, type.v.toNonNullable());
    if (this.ordered()) acc.ordered(true);
    if (this.caseInsensitive()) acc.caseInsensitive(true);
    Iterator it = pairsIterator();
    while (it.hasNext())
    {
      Entry e = (Entry)it.next();
      Object key = e.getKey();
      Object val = e.getValue();
      if (val != null)
        acc.set(key, val);
    }
    return acc;
  }

  public final Map<K,V> exclude(Func f)
  {
    Map acc = new Map(type);
    if (this.ordered()) acc.ordered(true);
    if (this.caseInsensitive()) acc.caseInsensitive(true);
    Iterator it = pairsIterator();
    while (it.hasNext())
    {
      Entry e = (Entry)it.next();
      Object key = e.getKey();
      Object val = e.getValue();
      if (!f.callBool(val, key))
        acc.set(key, val);
    }
    return acc;
  }

  public final boolean any(Func f)
  {
    if (map.size() == 0) return false;
    Iterator it = pairsIterator();
    while (it.hasNext())
    {
      Entry e = (Entry)it.next();
      Object key = e.getKey();
      Object val = e.getValue();
      if (f.callBool(val, key))
        return true;
    }
    return false;
  }

  public final boolean all(Func f)
  {
    if (map.size() == 0) return true;
    Iterator it = pairsIterator();
    while (it.hasNext())
    {
      Entry e = (Entry)it.next();
      Object key = e.getKey();
      Object val = e.getValue();
      if (!f.callBool(val, key))
        return false;
    }
    return true;
  }

  public final Object reduce(Object reduction, Func f)
  {
    Iterator it = pairsIterator();
    while (it.hasNext())
    {
      Entry e = (Entry)it.next();
      Object key = e.getKey();
      Object val = e.getValue();
      reduction = f.call(reduction, val, key);
    }
    return reduction;
  }

  public final Map map(Func f)
  {
    Type r = f.returns();
    if (r == Sys.VoidType) r = Sys.ObjType.toNullable();
    Map acc = new Map(type.k, r);
    if (this.ordered()) acc.ordered(true);
    if (this.caseInsensitive()) acc.caseInsensitive(true);
    Iterator it = pairsIterator();
    while (it.hasNext())
    {
      Entry e = (Entry)it.next();
      Object key = e.getKey();
      Object val = e.getValue();
      acc.set(key, f.call(val, key));
    }
    return acc;
  }

  public final Map mapNotNull(Func f)
  {
    Type r = f.returns();
    if (r == Sys.VoidType) r = Sys.ObjType;
    Map acc = new Map(type.k, r.toNonNullable());
    if (this.ordered()) acc.ordered(true);
    if (this.caseInsensitive()) acc.caseInsensitive(true);
    Iterator it = pairsIterator();
    while (it.hasNext())
    {
      Entry e = (Entry)it.next();
      Object key = e.getKey();
      Object val = e.getValue();
      acc.addNotNull(key, f.call(val, key));
    }
    return acc;
  }

//////////////////////////////////////////////////////////////////////////
// Conversion
//////////////////////////////////////////////////////////////////////////

  public final String join(String sep) { return join(sep, null); }
  public final String join(String sep, Func f)
  {
    int size = size();
    if (size == 0) return "";
    StringBuilder s = new StringBuilder(32+size*32);
    Iterator it = pairsIterator();
    while (it.hasNext())
    {
      Entry e = (Entry)it.next();
      Object key = e.getKey();
      Object val = e.getValue();
      if (s.length() > 0) s.append(sep);
      if (f == null)
        s.append(key).append(": ").append(val);
      else
        s.append(f.call(val, key));
    }
    return s.toString();
  }

  public final String toCode()
  {
    int size = size();
    StringBuilder s = new StringBuilder(32+size*32);
    s.append(type.signature());
    s.append('[');
    if (size == 0) s.append(':');
    Iterator it = pairsIterator();
    boolean first = true;
    while (it.hasNext())
    {
      Entry e = (Entry)it.next();
      Object key = e.getKey();
      Object val = e.getValue();
      if (first) first = false;
      else s.append(',').append(' ');
      s.append(FanObj.trap(key, "toCode", null))
       .append(':')
       .append(FanObj.trap(val, "toCode", null));
    }
    s.append(']');
    return s.toString();
  }

//////////////////////////////////////////////////////////////////////////
// Readonly
//////////////////////////////////////////////////////////////////////////

  public final boolean isRW()
  {
    return !readonly;
  }

  public final boolean isRO()
  {
    return readonly;
  }

  public final Map<K,V> rw()
  {
    if (!readonly) return this;

    Map rw = new Map(type);
    rw.map = (HashMap)map.clone();
    rw.readonly = false;
    rw.readonlyMap = this;
    rw.def = def;
    return rw;
  }

  public final Map<K,V> ro()
  {
    if (readonly) return this;
    if (readonlyMap == null)
    {
      Map ro = new Map(type);
      ro.map = map;
      ro.def = def;
      ro.readonly = true;
      readonlyMap = ro;
    }
    return readonlyMap;
  }

  public final boolean isImmutable()
  {
    return immutable;
  }

  public final Object toImmutable()
  {
    if (immutable) return this;

    // allocate new map of correct type
    HashMap temp;
    if (caseInsensitive()) temp = new CIHashMap(map.size()*2+3);
    else if (ordered()) temp = new LinkedHashMap(map.size()*2+3);
    else temp = new HashMap(map.size()*2+3);

    // make safe copy
    Iterator it = pairsIterator();
    while (it.hasNext())
    {
      Entry e = (Entry)it.next();
      Object key = e.getKey();
      Object val = e.getValue();

      if (val != null)
        val = FanObj.toImmutable(val);

      temp.put(key, val);
    }

    // return new immutable map
    Map ro = new Map(type, temp);
    ro.readonly = true;
    ro.immutable = true;
    ro.def = def;
    return ro;
  }

  private void modify()
  {
    // if readonly then throw readonly exception
    if (readonly)
      throw ReadonlyErr.make("Map is readonly");

    // if we have a cached readonlyMap, then detach
    // it so it remains immutable
    if (readonlyMap != null)
    {
      readonlyMap.map = (HashMap)map.clone();
      readonlyMap = null;
    }
  }

//////////////////////////////////////////////////////////////////////////
// Java
//////////////////////////////////////////////////////////////////////////

  public int sz() { return map.size(); }

  public Iterator pairsIterator()
  {
    if (map instanceof CIHashMap)
      return ((CIHashMap)map).pairs().iterator();
    else
      return map.entrySet().iterator();
  }

  public Iterator keysIterator()
  {
    return map.keySet().iterator();
  }

  /** Direct access to underlying hashmap */
  public HashMap toJava()
  {
    modify();
    return map;
  }

//////////////////////////////////////////////////////////////////////////
// CIHashMap (Case Insensitive)
//////////////////////////////////////////////////////////////////////////

  static class CIHashMap extends HashMap
  {
    public CIHashMap() {}
    public CIHashMap(int capacity) { super(capacity); }
    public Object get(Object key) { return super.get(new CIKey((String)key)); }
    public boolean containsKey(Object key) { return super.containsKey(new CIKey((String)key)); }
    public Object put(Object key, Object val) { return super.put(new CIKey((String)key), val); }
    public Object remove(Object key) { return super.remove(new CIKey((String)key)); }
    public Set pairs() { return new CIPairs(entrySet()); }

    public Set keySet()
    {
      java.util.HashSet keys = new java.util.HashSet();
      Iterator it = pairs().iterator();
      while (it.hasNext())
      {
        CIEntry entry = (CIEntry)it.next();
        keys.add(entry.key);
      }
      return keys;
    }

    public int hashCode()
    {
      int hash = 0;
      Iterator it = pairs().iterator();
      while (it.hasNext())
        hash += it.next().hashCode();
      return hash;
    }

    public boolean equals(Object obj)
    {
      if (!(obj instanceof HashMap)) return false;
      HashMap that = (HashMap)obj;
      if (size() != that.size()) return false;
      Iterator it = pairs().iterator();
      while (it.hasNext())
      {
        CIEntry entry = (CIEntry)it.next();
        Object thatVal = that.get(entry.key);
        if (!OpUtil.compareEQ(entry.val, thatVal)) return false;
      }
      return true;
    }
  }

  static final class CIPairs extends AbstractSet
  {
    CIPairs(Set set) { this.set = set; }
    public int size() { return set.size(); }
    public Iterator iterator() { return new CIPairsIterator(set.iterator()); }
    Set set;
  }

  static final class CIPairsIterator implements Iterator
  {
    CIPairsIterator(Iterator it) { this.it = it; }
    public boolean hasNext() { return it.hasNext(); }
    public Object next() { entry.set((Entry)it.next()); return entry; }
    public void remove() { it.remove(); }
    Iterator it;
    CIEntry entry = new CIEntry();
  }

  static final class CIEntry implements Entry
  {
    public void set(Entry e) { key = ((CIKey)e.getKey()).key; val = e.getValue(); }
    public Object getKey() { return key; }
    public Object getValue() { return val; }
    public int hashCode() { return key.hashCode() ^ (val == null ? 0 : val.hashCode()); }
    public boolean equals(Object o) { throw new UnsupportedOperationException(); }
    public Object setValue(Object v) { throw new UnsupportedOperationException(); }
    String key;
    Object val;
  }

  static final class CIKey
  {
    CIKey(String key) { this.key = key; this.hash = FanStr.caseInsensitiveHash(key); }
    public final int hashCode() { return hash; }
    public final boolean equals(Object obj) { return FanStr.equalsIgnoreCase(key, ((CIKey)obj).key); }
    public final String toString() { return key; }
    final String key;
    final int hash;
  }

//////////////////////////////////////////////////////////////////////////
// java.util.Map
//////////////////////////////////////////////////////////////////////////

  public final int size()
  {
    return map.size();
  }

  public boolean containsValue(Object val)
  {
    return map.containsValue(val);
  }

  public V put(K key, V val)
  {
    V old = get(key);
    set(key, val);
    return old;
  }

  public void putAll(java.util.Map<? extends K,? extends V> m)
  {
    for (Map.Entry<K,V> e : map.entrySet())
    {
      set(e.getKey(), e.getValue());
    }
  }

  public final void clear()
  {
    _clear();
  }

  public Collection<V> values()
  {
    return map.values();
  }

  public Set<K> keySet()
  {
    return map.keySet();
  }

  public Set<Map.Entry<K,V>> entrySet()
  {
    return map.entrySet();
  }

//////////////////////////////////////////////////////////////////////////
// Fields
//////////////////////////////////////////////////////////////////////////

  private MapType type;
  private HashMap<K,V> map;
  private Map<K,V> readonlyMap;
  private boolean readonly;
  private boolean immutable;
  private V def;

}

